lib/parsers/iso_base_media_file_format/decoder.rb

```diff
@@ -0,0 +1,1039 @@
+
+      # A mapping of atom types to their respective parser methods. Each method must take a single Integer parameter, size,
+      # and return the atom's fields and children where appropriate as a Hash and Array of Atoms respectively.
+      ATOM_PARSERS = {
+        # 'bxml' => :bxml,
+        # 'co64' => :co64,
+        # 'cprt' => :cprt,
+        # 'cslg' => :cslg,
+        # 'ctts' => :ctts,
+        'dinf' => :container,
+        # 'dref' => :dref,
+        'edts' => :container,
+        # 'fecr' => :fecr,
+        # 'fiin' => :fiin,
+        # 'fire' => :fire,
+        # 'fpar' => :fpar,
+        # 'ftyp' => :typ,
+        # 'gitn' => :gitn,
+        # 'hdlr' => :hdlr,
+        # 'hmhd' => :hmhd,
+        # 'iinf' => :iinf,
+        # 'iloc' => :iloc,
+        # 'infe' => :infe,
+        # 'ipro' => :ipro,
+        # 'iref' => :iref,
+        # 'leva' => :leva,
+        # 'mdhd' => :mdhd,
+        'mdia' => :container,
+        'meco' => :container,
+        # 'mehd' => :mehd,
+        # 'mere' => :mere,
+        # 'meta' => :meta,
+        # 'mfhd' => :mfhd,
+        'mfra' => :container,
+        # 'mfro' => :mfro,
+        'minf' => :container,
+        'moof' => :container,
+        'moov' => :container,
+        'mvex' => :container,
+        # 'mvhd' => :mvhd,
+        'nmhd' => :empty,
+        # 'padb' => :padb,
+        'paen' => :container,
+        # 'pdin' => :pdin,
+        # 'pitm' => :pitm,
+        # 'prft' => :prft,
+        # 'saio' => :saio,
+        # 'saiz' => :saiz,
+        # 'sbgp' => :sbgp,
+        'schi' => :container,
+        # 'schm' => :schm,
+        # 'sdtp' => :sdtp,
+        # 'segr' => :segr,
+        # 'sgpd' => :sgpd,
+        # 'sidx' => :sidx,
+        'sinf' => :container,
+        # 'smhd' => :smhd,
+        # 'ssix' => :ssix,
+        'stbl' => :container,
+        # 'stco' => :stco,
+        # 'stdp' => :stdp,
+        'sthd' => :empty,
+        'strd' => :container,
+        # 'stri' => :stri,
+        'strk' => :container,
+        # 'stsc' => :stsc,
+        # 'stsd' => :stsd,
+        # 'stsh' => :stsh,
+        # 'stss' => :stss,
+        # 'stsz' => :stsz,
+        # 'stts' => :stts,
+        # 'styp' => :typ,
+        # 'stz2' => :stz2,
+        # 'subs' => :subs,
+        # 'tfra' => :tfra,
+        # 'tkhd' => :tkhd,
+        'trak' => :container,
+        # 'trex' => :trex,
+        # 'tsel' => :tsel,
+        'udta' => :container,
+        # 'url ' => :dref_url,
+        # 'urn ' => :dref_urn,
+        'uuid' => :uuid,
+        # 'vmhd' => :vmhd,
+        # 'xml ' => :xml,
+      }
+
+      # Parse the atom at the IO's current position.
+      #
+      # @return [Atom, nil]
+      def parse_atom
+        position = @buf.pos
+
+        size = read_int_32
+        type = read_string(4)
+        size = read_int_64 if size == 1
+        body_size = size - (@buf.pos - position)
+        next_atom_position = position + size
+
+        if self.class::ATOM_PARSERS.include?(type)
```

Comment:
Why not just:

```suggestion
        if ATOM_PARSERS.include?(type)
```

Reply:
Doing it this way means that it continues to function with extending class (such as the CR3 decoder). This way, all that extending classes have to do is add to its own ATOM_PARSERS, like [so](https://github.com/WeTransfer/format_parser/pull/222/files#diff-8e8d366860cab7abfad65caebdb7891360815ec2013a1cad34d4efdb7ee81c06R8-R10).

Reply:
Ah, got it. It is a bit subtle, seems more suitable to leverage inheritance here. But I'll leave it up to you.

Reply:
Tried doing it with standard inheritance, but I think I'm too much of a ruby n00b and this was what I found online as the suggested way to modify an inherited constant and have it be picked up by an inherited method without unintended changes to the base class or other inheriting classes. Played around with it a good bit and this was the way that worked. If you've got a suggestion for how to make this more intuitive then I'm definitely all ears!

Reply:
We can have a look together next week, but this works as well, so don't let it block you 😊